docs/architecture.md

```diff
@@ -0,0 +1,126 @@
+# Architecture of a consensus node
+
+
+This is a block diagram of all the parts of the system and will be updated as needed.
+
+```mermaid
+graph LR
+subgraph "Consensus Node"
+  engine[Engine API Client]
+  BAPI[Beacon API]
+  TICK[Slot processor]
+	blk_db[Block DB]
+	BS_db[Beacon State DB]
+  brod[Broadway]
+	FCTree[Fork choice store - Genserver]
+  BAPI -->|Beacon state queries| BS_db
+	brod -->|Save blocks| blk_db
+	brod -->|Blocks and attestations| FCTree
+	TICK -->|New ticks| FCTree
+  BAPI --> engine
+	BAPI --> |head/slot requests| FCTree
+	brod --> |Save new states|BS_db
+end
+GOS[Gossip Protocols]
+exec[Execution Client]
+VALIDATOR[Validator]
+engine <--> |payload validation, execution| exec
+GOS -->|blocks, attestations| brod
+VALIDATOR --> BAPI
+```
+
+## Networking
+
+The main entry for new events is the gossip protocol, which is the way in which our consensus node communicates with other consensus nodes. This includes:
+
+1. Discovery: our node has a series of known `bootnodes` hardcoded. We request them a list of the nodes the know about and we add them to our list. We save them locally and now can use those too to request new nodes.
+2. Message propagation. When a proposer sends a new block, or validators attest for a new block, they sent those to other known nodes. Those, in turn propagate the messages sent to other nodes. This process is repeated until, ideally, the whole network receives the messages.
+
+We use a library called `libp2p`, implemented in `go`, for the networking primitives.
```

Comment:
This is a bit misleading. To clarify, and quoting [the libp2p repo's readme](https://github.com/libp2p/libp2p):

> libp2p is a modular networking framework bundled together as a full stack of protocols for peer-to-peer systems.

The actual implementation we are using is `go-libp2p` (+ other `go-libp2p-*` modules), and the discovery part of `go-ethereum`'s P2P stack.

I'd rephrase this as:

```
We use the `go-libp2p` library for the networking primitives, which is an implementation of the `libp2p`networking stack.
```